esp-hal/src/aes/mod.rs

```diff
@@ -595,6 +751,480 @@ pub mod dma {
+        ///
+        /// ## Example
+        ///
+        /// ```rust, no_run
+        /// # {before_snippet}
+        /// use esp_hal::aes::dma::AesDmaBackend;
+        ///
+        /// # {dma_channel}
+        /// let mut aes = AesDmaBackend::new(peripherals.AES, dma_channel);
+        /// let _handle = aes.start();
+        /// # {after_snippet}
+        /// ```
+        pub fn start(&mut self) -> AesDmaWorkQueueDriver<'_, 'd> {
+            AesDmaWorkQueueDriver {
+                _inner: WorkQueueDriver::new(self, AES_DMA_VTABLE, &super::AES_WORK_QUEUE),
+            }
+        }
+
+        // WorkQueue callbacks. They may run in any context.
+
+        unsafe fn from_raw<'any>(ptr: *const ()) -> &'any mut Self {
+            unsafe { unwrap!(ptr.cast_mut().cast::<AesDmaBackend<'_>>().as_mut()) }
+        }
+
+        fn start_processing(&mut self, item: &mut AesOperation) {
+            let driver = unwrap!(self.driver.as_mut());
+
+            // There are some constraints that make us (partially) fall back to CPU-driven mode:
+            // - The algo isn't implemented in hardware. Not much to do here, we process the whole
+            //   data using the CPU.
+            // - There is data stuck in the cipher state.
+            //   - In this case we need to flush the data before we can start using the DMA.
+            // - The data alignment isn't appropriate to its location (PSRAM)
+            //   - The DMA can read (transmit) any number of bytes from any address (in theory), but
+            //     it can't write arbitrarily to PSRAM. We should be able to split the write
+            //     transfer into two parts, and write the unaligned bytes into internal memory using
+            //     the DMA, then copy the data out to PSRAM using the CPU at the end.
+            // - The data length (after all the above) is not a multiple of the block length
+            //   - We process as many blocks as we can using the DMA, then use the CPU for the
+            //     remainder.
+
+            if unsafe { !item.cipher_mode.as_ref().implemented_in_hardware() } {
+                // Algo is either not implemented in hardware, or the data is not accessible by DMA.
+                driver.aes.process_work_item(item);
+                return;
+            }
+
+            if !crate::soc::is_valid_memory_address(item.buffers.input.addr().get()) {
+                unsafe {
+                    // Safety:
+                    // We've verified when constructing the buffer that input and output are the
+                    // same length. The CryptoBuffers constructor also ensures that output is
+                    // mutable and therefore is in RAM. If input is not, the pointers
+                    // don't overlap.
+                    item.buffers.output.cast::<u8>().copy_from_nonoverlapping(
+                        item.buffers.input.cast::<u8>(),
+                        item.buffers.input.len(),
+                    );
```

Comment:
What about `CryptoBuffers::new_in_place`? It allows the pointers to overlap.

This safety comment should also mention how it is enforced that the `new_in_place` c'tor wasn't used.